### DISCUSSION TITLE
checkout not checking out branch

### DISCUSSION THREAD

**Discussion by ghost** [actions/checkout#1637] [General]
Is anyone else having issues with this, the README says

    # The branch, tag or SHA to checkout. When checking out the repository that
    # triggered a workflow, this defaults to the reference or SHA for that event.
    # Otherwise, uses the default branch.
    ref: ''

    1. Created PR
    2. Checked in the PR
    3. Build is kicked off
    4. Have code as below, but when I print branch it prints the main branch, not the branch that created the pull_request the triggered the workflow.

    name: Preflight Job
permissions:
  id-token: write
  contents: write
on:
  pull_request:
    types: [opened, reopened]

  issue_comment:
    types: [ created ]

  workflow_dispatch:

jobs:
  preflight-vars:
    if: github.event.issue.pull_request && contains(github.event.comment.body, '/test')
    runs-on: [ HOST ]

    steps:

      - name: Checkout code
        uses: actions/checkout@v4
        with:
          fetch-depth: 0

      - name: Print Branch
        run: |
          echo "Print Branch"
          echo "$(git branch)"
          
          The branch printed is always the default branch, not the branch the created the event that triggered the workflow

**Comment by ghost** [comment]
NOTE: We discovered this is working correctly, we discovered that adding a conditional such as: if: github.event.issue.pull_request && contains(github.event.comment.body, 'test this please') then the workflow runs against the default branch, and not the branch that kicked off the workflow.  If I remove the conditional all together, then the workflow runs against the PR branch.

I think per this url: https://docs.github.com/en/actions/learn-github-actions/contexts that the 'runner' is changing the $GITHUB_REF to default branch. 

In this example, the if statement checks the github.ref context to determine the current branch name; if the name is refs/heads/main, then the subsequent steps are executed. The if check is processed by GitHub Actions, and the job is only sent to the runner if the result is true. Once the job is sent to the runner, the step is executed and refers to the $GITHUB_REF variable from the runner.

So we think the runner is changing to the default branch.  We are working with the implementer of the runner to troubleshoot.

**Comment by metmikel** [comment]
I think the cause of this might be that the action is triggered by `pull_request` events.  When a `pull_request` event is used to trigger the workflow, the `github.ref` is set to the reference of the merge branch (which includes the history of the both the branch being merged from and that being merged to) rather than the branch which is being merged from (the branch you pushed to).  See the documentation at:

https://docs.github.com/en/actions/writing-workflows/choosing-when-your-workflow-runs/events-that-trigger-workflows#pull_request

The relevant bit being:

> Note that GITHUB_SHA for this event is the last merge commit of the pull request merge branch. If you want to get the commit ID for the last commit to the head branch of the pull request, use github.event.pull_request.head.sha instead.

If you are using other types of events to trigger the workflow as well, then you will need to add a conditional statement to the `ref` parameter to deal with both cases, i.e.:

```
        uses: actions/checkout@v4
        with:
          ref: ${{ github.event_name == 'pull_request' && github.event.pull_request.head.ref || github.ref }}
```

**Reply by poly2it** [reply]
THANK YOU!
